Internationalized docs (#2772)
adobe/react-spectrum: packages/@internationalized/date/docs/ZonedDateTime.mdx

This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import docs from 'docs:@internationalized/date';
import {HeaderInfo, FunctionAPI, ClassAPI, TypeContext, InterfaceType, TypeLink, PageDescription} from '@react-spectrum/docs';
import packageData from '@internationalized/date/package.json';

---
category: Date and Time
keywords: [date, time, timezone, calendar, internationalization]
after_version: 3.0.0
---

# ZonedDateTime

<PageDescription>{docs.exports.ZonedDateTime.description}</PageDescription>

<HeaderInfo
  packageData={packageData}
  componentNames={['ZonedDateTime']}
  sourceData={[]} />

## Introduction

> A `CalendarDateTime` object represents an exact date and time in a specific time zone, in a specific calendar system such as the Gregorian calendar. Use this type to represent an exact moment in time at a particular location on Earth.
> 
> A `CalendarDateTime` can be created using the constructor. This example creates a date that represents February 3rd, 2022 at 9:15 AM in the Gregorian calendar system.
> 
> ```tsx
> import {CalendarDateTime} from '@internationalized/date';

Reviewer: ```suggestion
A `ZonedDateTime ` object represents an exact date and time in a specific time zone, in a specific calendar system such as the Gregorian calendar. Use this type to represent an exact moment in time at a particular location on Earth.

A `ZonedDateTime ` can be created using the constructor. This example creates a date that represents February 3rd, 2022 at 9:15 AM in the Gregorian calendar system with a time zone of `"America/Los Angeles"`.

```tsx
import {ZonedDateTime} from '@internationalized/date';
```

let date = new ZonedDateTime(
  // Date
  2022, 2, 3,

>   // Time zone and UTC offsset

Reviewer: mega nit
```suggestion
  // Time zone and UTC offset
```

  'America/Los_Angeles', -28800000,
  // Time
  12, 24, 45
);
```

> You can also create a `CalendarDate` by parsing an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) formatted string using the one of the following functions:

Reviewer: ```suggestion
You can also create a `ZonedDateTime` by parsing an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) formatted string using the one of the following functions:
```

* <TypeLink links={docs.links} type={docs.exports.parseAbsolute} /> – This function parses an absolute date and time that occurs at the same instant at all locations on Earth. It can be represented in UTC (e.g. `"2021-11-07T07:45:00Z"`), or stored with a particular offset (e.g. `"2021-11-07T07:45:00-07:00"`). A time zone identifier, e.g. `America/Los_Angeles`, must be passed, and the result will be converted into that time zone. Absolute times are the best way to represent events that occurred in the past, or future events where an exact time regardless of time zone is needed.
* <TypeLink links={docs.links} type={docs.exports.parseAbsoluteToLocal} /> – This function parses an absolute date and time into the current user's local time zone. It is a shortcut for `parseAbsolute`, and accepts the same formats.

```tsx
import {parseZonedDateTime} from '@internationalized/date';

let date = parseZonedDateTime('2022-11-07T00:45[America/Los_Angeles]');
let date = parseAbsolute('2021-11-07T07:45:00Z', 'America/Los_Angeles');
let date = parseAbsoluteToLocal('2021-11-07T07:45:00Z');
```

> Once you have a `ZonedDateTime` object, you can read its properties, or manipulate it as described in the [Manipulating dates](#manipulating-dates) section below. You can also convert it to an ISO 8601 string, native JavaScript `Date` object, or another representation. See the [Conversion](#conversion) section below for details.

Reviewer 2: Should this `ISO 8601` be a link?

Author: Seems unnecessary to repeat since there is already a link on the first instance above?

```tsx
let date = new ZonedDateTime(
  2022, 2, 3,
  'America/Los_Angeles', -28800000,
  12, 24, 45
);

date.toString(); // '2022-02-03T12:24:45-08:00[America/Los_Angeles]'
```

## Interface

<ClassAPI links={docs.links} class={docs.exports.ZonedDateTime} />

## Calendar systems

By default, `ZonedDateTime` uses the Gregorian calendar system, but many other calendar systems that are used around the world are supported, such as Hebrew, Indian, Islamic, Buddhist, Ethiopic, and more. A <TypeLink links={docs.links} type={docs.exports.Calendar} /> instance can be passed to the `ZonedDateTime` constructor to represent dates in that calendar system.

This example creates a date in the Buddhist calendar system, which is equivalent to April 4th, 2020 at 9:15 AM in the Gregorian calendar.

```tsx
import {BuddhistCalendar} from '@internationalized/date';

let date = new ZonedDateTime(
  new BuddhistCalendar(), 2563, 4, 30,
  'America/Los_Angeles', -28800000,
  9, 15
);
```

> See the [Calendar](Calendar.html) docs for details about the supported calendars.

Reviewer: ```suggestion
See the [Calendar](Calendar.html#implementations) docs for details about the supported calendars.
```

### Eras

> Many calendar systems have only one era, or a modern era and a pre-modern era (e.g. AD and BC in the Gregorian calendar). However, other calendar systems may have many eras. For example, the Japanese calendar has eras for the reign of each Emperor. `CalendarDateTime` represents eras using string identifiers, which can be passed as an additional parameter to the constructor before the year. When eras are present, years are numbered starting from 1 within the era.

Reviewer: ```suggestion
Many calendar systems have only one era, or a modern era and a pre-modern era (e.g. AD and BC in the Gregorian calendar). However, other calendar systems may have many eras. For example, the Japanese calendar has eras for the reign of each Emperor. `ZonedDateTime` represents eras using string identifiers, which can be passed as an additional parameter to the constructor before the year. When eras are present, years are numbered starting from 1 within the era.
```

This example creates a date in the Japanese calendar system, which is equivalent to April 4th, 2020 at 9:15 AM in the Gregorian calendar.

```tsx
import {JapaneseCalendar} from '@internationalized/date';

let date = new ZonedDateTime(
  new JapaneseCalendar(), 'heisei', 31, 4, 30,
  'America/Los_Angeles', -28800000,
  9, 15
);
```

A list of valid era identifiers can be retried using the `getEras` method of a `Calendar` instance. If an era is not provided, the date is assumed to be in the current era.

### Converting between calendars

The <TypeLink links={docs.links} type={docs.exports.toCalendar} /> function can be used to convert a date from one calendar system to another.

This example converts a Gregorian date to a Hebrew one.

```tsx
import {toCalendar, HebrewCalendar} from '@internationalized/date';

let gregorianDate = new ZonedDateTime(
  2020, 9, 19,
  'America/Los_Angeles', -28800000,
  10, 30
);

let hebrewDate = toCalendar(gregorianDate, new HebrewCalendar());
```

## Manipulating dates

### Adding and subtracting durations

> A <TypeLink links={docs.links} type={docs.exports.DateTimeDuration} /> is an object that represents an amount of time, with fields such as `years`, `months`, `hours`, and `minutes`. The `add` and `subtract` methods of `CalendarDateTime` objects can be used to adjust the date by the given duration. These methods return a new date, and do not mutate the original.

Reviewer: ```suggestion
A <TypeLink links={docs.links} type={docs.exports.DateTimeDuration} /> is an object that represents an amount of time, with fields such as `years`, `months`, `hours`, and `minutes`. The `add` and `subtract` methods of `ZonedDateTime` objects can be used to adjust the date by the given duration. These methods return a new date, and do not mutate the original.
```

```tsx
let date = parseZonedDateTime('2022-02-03T09:45[America/Los_Angeles]');

> date.add({weeks: 1}); // 2022-02-09T09:45[America/Los_Angeles]

Reviewer 2: ```suggestion
date.add({weeks: 1}); // 2022-02-10T09:45[America/Los_Angeles]
```

date.add({months: 1}); // 2022-03-03T09:45[America/Los_Angeles]
date.add({years: 1, months: 1, days: 1}); // 2023-03-04T09:45[America/Los_Angeles]
date.add({hours: 1}); // 2022-02-03T10:45[America/Los_Angeles]
date.add({minutes: 30}); // 2022-02-03T10:15[America/Los_Angeles]

date.subtract({weeks: 1}); // 2022-01-27T09:45[America/Los_Angeles]
date.subtract({months: 1}); // 2022-01-03T09:45[America/Los_Angeles]
date.subtract({years: 1, months: 1, days: 1}); // 2021-01-02T09:45[America/Los_Angeles]
date.subtract({hours: 1}); // 2022-02-03T8:45[America/Los_Angeles]
date.subtract({minutes: 30}); // 2022-02-03T09:15[America/Los_Angeles]
```

> Adding or subtracting a duration that goes beyond the limits of a particular field will cause the date to be _balanced_. For example, adding one day to August 31st results in September 1st. In addition, if adding or subtracting one field causes another to be invalid, the date will be _constrained_. For example, adding one month to August 31st results in September 30th because September 31st does not exist.

Reviewer 2: I'm looking at the next paragraph and wondering if we should use quotes instead of italics. I'm concerned about another formatting style to "confuse" readers and another rule for us to follow.

* `'reject'` – throws an error when the time is ambiguous

```tsx
// A "spring forward" transition
let date = parseZonedDateTime('2020-03-01T02:00-08:00[America/Los_Angeles]');
date.set({day: 8}); // 2020-03-08T03:00:00-07:00[America/Los_Angeles]
date.set({day: 8}, 'earlier'); // 2020-03-08T01:00:00-08:00[America/Los_Angeles]
date.set({day: 8}, 'later'); // 2020-03-08T03:00:00-07:00[America/Los_Angeles]

// A "fall back" transition
let date = new CalendarDateTime(2020, 11, 1, 1);
let date = parseZonedDateTime('2020-10-01T01:00-07:00[America/Los_Angeles]');
date.set({month: 11}); // 2020-11-01T01:00:00-07:00[America/Los_Angeles]
date.set({month: 11}, 'earlier'); // 2020-11-01T01:00:00-07:00[America/Los_Angeles]
date.set({month: 11}, 'later'); // 2020-11-01T01:00:00-08:00[America/Los_Angeles]
```

### Cycling fields

The `cycle` method allows incrementing or decrementing a single field. It is similar to the `add` and `subtract` methods, but when the value reaches the minimum or maximum, it wraps around rather than affecting other fields.

```tsx
let date = parseZonedDateTime('2022-12-31T23:59[America/Los_Angeles]');

> date.cycle('day', 1); // 2022-12-01T23:59
> date.cycle('month', 1); // 2022-01-31T23:59
> date.cycle('hour', 1); // 2022-12-31T00:59
> date.cycle('minute', 1); // 2022-12-31T23:00

Reviewer: Nit:
```suggestion
date.cycle('day', 1); // 2022-12-01T23:59[America/Los_Angeles]
date.cycle('month', 1); // 2022-01-31T23:59[America/Los_Angeles]
date.cycle('hour', 1); // 2022-12-31T00:59[America/Los_Angeles]
date.cycle('minute', 1); // 2022-12-31T23:00[America/Los_Angeles]
```

a.compare(b) < 0; // true
b.compare(a) > 0; // true
```

In addition, the following functions can be used to perform a partial comparison. These functions accept dates in different calendar systems, and the second date is converted to the calendar system of the first date before comparison.

* <TypeLink links={docs.links} type={docs.exports.isSameYear} /> – <span>{docs.exports.isSameYear.description}</span>
* <TypeLink links={docs.links} type={docs.exports.isSameMonth} /> – <span>{docs.exports.isSameMonth.description}</span>
* <TypeLink links={docs.links} type={docs.exports.isSameDay} /> – <span>{docs.exports.isSameDay.description}</span>
* <TypeLink links={docs.links} type={docs.exports.isToday} /> – <span>{docs.exports.isToday.description}</span>

```tsx
import {isSameMonth, IslamicUmalquraCalendar} from '@internationalized/date';

isSameMonth(
  parseZonedDateTime('2022-04-16T09:45[America/Los_Angeles]'),
  parseZonedDateTime('2022-04-30T05:15[America/Los_Angeles]'),
); // true

isSameMonth(
  parseZonedDateTime('2022-04-16T09:45[America/Los_Angeles]'),
  parseZonedDateTime('2022-08-020T05:15[America/Los_Angeles]'),

> ); // false

Reviewer: Is this example supposed to construct a ZonedDateTime with the IslamicUmalquraCalendar? If not we can remove the import but I figure it is supposed to as to show the difference between `isSameMonth` and `isEqualMonth` when comparing dates in different calendar systems

Author: maybe originally. I think the constructor is too long with ZonedDateTime though, so I guess I simplified it and forgot to remove the import.

```

> A similar set of functions is also available that does not convert between calendar systems and requires the calendars to be equal.

Reviewer 2: In other places you have examples of `isEqualXXXX`.

Author: Yeah it's just harder to show because ZonedDateTime has so many arguments for the constructor.
